mingw: lstat: compute correct size for symlinks

billziss-gh opened the pull request:
This PR includes a single commit that fixes the `mingw_lstat` function to return correct size for symlinks on Windows. Prior to this commit the function returned a fixed size of 4096, which caused problems for repositories that were accessed using git from different platforms.

For example, if using both git for WSL and git for Windows the problem that I describe below could arise.

Consider a git repository accessed using git for WSL:

<img width="519" alt="wsl" src="https://user-images.githubusercontent.com/15330948/82990914-bf8df200-9fb1-11ea-8a94-6d9da2563433.png">

Now consider the same repository accessed using git for Windows:

<img width="520" alt="ps" src="https://user-images.githubusercontent.com/15330948/82991077-fb28bc00-9fb1-11ea-82c7-48a04922ca78.png">

Notice that `RelNotes`, which is a symbolic link, appears modified. This happened because `mingw_lstat` returned `st_size == 4096`, which then git for Windows compared to the index prepared by git for WSL and found a size mismatch.

Finally, using the commit in this PR, a patched `git-status.exe` reports correctly that there are no changes found.

<img width="457" alt="mingw" src="https://user-images.githubusercontent.com/15330948/82991571-b94c4580-9fb2-11ea-8dcf-ab2d03bc2f26.png">

dscho commented on compat/mingw.c:
```diff
@@ -3029,6 +3030,60 @@ int readlink(const char *path, char *buf, size_t bufsiz)
+	int len;
+
+	/* read reparse point data */
+	handle = CreateFileW(wpath, 0,
```
In our performance tests, `CreateFile()` was _substantially_ slower than `FindFirstFile()`. This might not sound like a big thing, but when you have tens of thousands of files (or even [several million files](https://devblogs.microsoft.com/bharry/the-largest-git-repo-on-the-planet/)), it does matter.

Noticing that symbolic links are relatively rare, would it not make more sense to replace the `MAX_LONG_PATH` constant by a function that determines the length of a symbolic link (and is called _only_ for symbolic links)?

dscho replied:
And while we're already at that point where we want to read the length of the symlink, would it not make more sense to use the [`readlink()`](https://github.com/git-for-windows/git/blob/v2.27.0-rc2.windows.1/compat/mingw.c#L2974-L3033) function rather than duplicating (part of) its logic?

billziss-gh replied:
Thank you for the quick review.

> In our performance tests, `CreateFile()` was substantially slower than `FindFirstFile()`.

The `FindFirstFileW`, `FindClose` sequence makes the following calls:

- `NtOpenFile` / `IoCreateFile` on the **parent directory** of the supplied file path.
- `NtQueryDirectoryFile` with the requested file name or mask of the supplied file path.
- `NtClose`

The `CreateFileW`, `DeviceIoControl`, `CloseHandle` sequence makes the following calls:

- `NtCreateFile` / `IoCreateFile` on the supplied file path.
- `NtFsControlFile` with `FSCTL_GET_REPARSE_POINT`.
- `NtClose`.

So if you have observed performance differences it is perhaps because when accessing multiple files in the same directory the file system has a better chance to cache directory contents in the `FindFirstFileW` sequence.

Even more importantly for our case, `FindFirstFileW` does not give us enough information to fill in the symlink `stat`. It gives us the reparse tag, but does not give us the length of the symlink. So querying the reparse point for its data and computing the length in that way becomes unavoidable.

> Noticing that symbolic links are relatively rare, would it not make more sense to replace the `MAX_LONG_PATH` constant by a function that determines the length of a symbolic link (and is called only for symbolic links)?

I agree. But please note that this happens already: the expensive reparse point computation happens only when the `if (fdata.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)` condition is true.

> And while we're already at that point where we want to read the length of the symlink, would it not make more sense to use the `readlink()` function rather than duplicating (part of) its logic?

I agree and that is how I implemented it at first. Unfortunately `readlink` does not return the actual reparse tag that is used in the computation of `st_mode` by `file_attr_to_st_mode`. Without that special `wreadlink_len` function we would have to issues both a `FindFirstFileW` and a `readlink` to properly `stat` a symlink.

billziss-gh replied:
BTW, one improvement in the submitted commit is the following:

Currently the `wreadlink_len` returns -1 it if encounters a reparse point that is not a symbolic link or a mount point; this causes `mingw_lstat` to also return -1. Perhaps such cases should be handled more gracefully (e.g. by returning the prior size of 4096 for compatibility purposes).

dscho replied:
> Without that special `wreadlink_len` function we would have to issue both a `FindFirstFileW` and a `readlink` to properly `stat` a symlink.

Well, in the current state, there is a lot of duplicated code.

The common strategy in Git's commit history is to refactor out a helper function that can be used from multiple call sites. In this case, it would be factoring out that part of `readlink()` that determines e.g. `wbuf` and `len`, then calling that from `readlink()` _and_ from `mingw_lstat()`.

That way, code is not duplicated, and as a consequence there is not even a possibility for one of the two copies of the code to become stale, as there is only a single copy of the code to begin with.

billziss-gh replied:
Understood. I will look into making the necessary changes and resubmit.

dscho replied:
Thank you!

billziss-gh commented:
I have made a couple of small updates in the commit:

- I renamed the internal function `wreadlink_len` to `get_reparse_point_info` to disambiguate it from `readlink`.

- The function returns the correct length of a symbolic link or mount point as before, but in addition it returns `MAX_LONG_PATH` for other reparse points (to be compatible with how git for Windows currently handles such reparse points).

Finally this commit has not been signed off.

dscho commented on compat/mingw.c:
```diff
@@ -969,10 +969,12 @@ static int has_valid_directory_prefix(wchar_t *wfilename)
 {
 	WIN32_FILE_ATTRIBUTE_DATA fdata;
-	WIN32_FIND_DATAW findbuf = { 0 };
+	DWORD reparse_tag;
```
You will need to initialize `reparse_tag` so that it can be passed to `file_attr_to_st_mode()` even if the path is not a reparse point.

billziss-gh replied:
Argh! You are correct of course.

dscho replied:
:-) No worries. I always hope that I can help improve the patches through my reviews.

billziss-gh replied:
Thanks. Somehow the compiler did not catch this, but I should have been more careful.

billziss-gh commented:
I have refactored the code as per your request:

- New `get_reparse_point` function can be used to get a reparse point.
- New `get_reparse_point_link_len` is used to get the reparse tag and link length for `mingw_lstat`.
- The `readlink` function now uses `get_reparse_point`.

dscho commented on compat/mingw.c:
```diff
@@ -3012,6 +3007,49 @@ int readlink(const char *path, char *buf, size_t bufsiz)
+		return MAX_LONG_PATH; /* return value for compatibility */
+	}
+
+	len = WideCharToMultiByte(CP_UTF8, 0, normalize_ntpath(wbuf), -1, 0, 0, NULL, NULL);
```
I would prefer this conversion from `wbuf` to UTF-8 to be identical to the `readlink()` case. Otherwise, it will be too easy to change one and forget the other, and then we end up with inconsistent lengths again.

Granted, `readlink()` uses

```c
	if ((len = xwcstoutf(tmpbuf, normalize_ntpath(wbuf), MAX_LONG_PATH)) < 0)
```

which we do not use here because we do not _really_ want to convert `wbuf` to UTF-8 at this point. But then, maybe we just shouldn't care _so_ much and use a temporary buffer in `get_reparse_point_link_len()` that is simply ignored?

If we do that, we can refactor pretty much everything in `readlink()` (except the initial `xutftowcs_long_path()` call) into a new function `int readlink_1(wchar_t *wpath, char *buf, size_t bufsiz, size_t *plen, DWORD *ptag)`. And then we do not even need a wrapper around _that_ when calling it in `mingw_lstat()`, and `readlink()` itself will be a very minimal wrapper that only converts `path` to `wpath` and then calls `return readlink_1(...);`.

dscho replied:
In other words, I would prefer this fixup squashed in:

```diff
diff --git a/compat/mingw.c b/compat/mingw.c
index 92db342833c8..87f985c6e7e9 100644
--- a/compat/mingw.c
+++ b/compat/mingw.c
@@ -969,7 +969,8 @@ static int has_valid_directory_prefix(wchar_t *wfilename)
 	return 1;
 }
 
-static int get_reparse_point_link_len(const WCHAR *wpath, DWORD *ptag);
+static int readlink_1(const WCHAR *wpath, char *tmpbuf, int *plen, DWORD *ptag);
+
 int mingw_lstat(const char *file_name, struct stat *buf)
 {
 	WIN32_FILE_ATTRIBUTE_DATA fdata;
@@ -989,10 +990,12 @@ int mingw_lstat(const char *file_name, struct stat *buf)
 	}
 
 	if (GetFileAttributesExW(wfilename, GetFileExInfoStandard, &fdata)) {
-		/* for reparse points, use get_reparse_point_link_len to get the link tag and length */
+		/* for reparse points, get the link tag and length */
 		if (fdata.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
-			link_len = get_reparse_point_link_len(wfilename, &reparse_tag);
-			if (link_len < 0)
+			char tmpbuf[MAX_LONG_PATH];
+
+			if (readlink_1(wfilename, tmpbuf, &link_len,
+				       &reparse_tag) < 0)
 				return -1;
 		}
 		buf->st_ino = 0;
@@ -2969,10 +2972,11 @@ typedef struct _REPARSE_DATA_BUFFER {
 } REPARSE_DATA_BUFFER, *PREPARSE_DATA_BUFFER;
 #endif
 
-static int get_reparse_point(const WCHAR *wpath, REPARSE_DATA_BUFFER *b, size_t siz, WCHAR **pwbuf)
+static int readlink_1(const WCHAR *wpath, char *tmpbuf, int *plen, DWORD *ptag)
 {
 	HANDLE handle;
 	WCHAR *wbuf;
+	REPARSE_DATA_BUFFER *b = alloca(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
 	DWORD dummy;
 
 	/* read reparse point data */
@@ -2985,7 +2989,7 @@ static int get_reparse_point(const WCHAR *wpath, REPARSE_DATA_BUFFER *b, size_t
 		return -1;
 	}
 	if (!DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, NULL, 0, b,
-			siz, &dummy, NULL)) {
+			MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &dummy, NULL)) {
 		errno = err_win_to_posix(GetLastError());
 		CloseHandle(handle);
 		return -1;
@@ -2993,7 +2997,7 @@ static int get_reparse_point(const WCHAR *wpath, REPARSE_DATA_BUFFER *b, size_t
 	CloseHandle(handle);
 
 	/* get target path for symlinks or mount points (aka 'junctions') */
-	switch (b->ReparseTag) {
+	switch ((*ptag = b->ReparseTag)) {
 	case IO_REPARSE_TAG_SYMLINK:
 		wbuf = (WCHAR*) (((char*) b->SymbolicLinkReparseBuffer.PathBuffer)
 				+ b->SymbolicLinkReparseBuffer.SubstituteNameOffset);
@@ -3007,52 +3011,28 @@ static int get_reparse_point(const WCHAR *wpath, REPARSE_DATA_BUFFER *b, size_t
 				+ b->MountPointReparseBuffer.SubstituteNameLength) = 0;
 		break;
 	default:
-		wbuf = NULL;
-		break;
-	}
-
-	*pwbuf = wbuf;
-	return 0;
-}
-
-static int get_reparse_point_link_len(const WCHAR *wpath, DWORD *ptag)
-{
-	WCHAR *wbuf;
-	REPARSE_DATA_BUFFER *b = alloca(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
-	int len;
-
-	if (get_reparse_point(wpath, b, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &wbuf) < 0)
+		errno = EINVAL;
 		return -1;
-	if (wbuf == NULL) {
-		*ptag = b->ReparseTag;
-		return MAX_LONG_PATH; /* return value for compatibility */
 	}
 
-	len = WideCharToMultiByte(CP_UTF8, 0, normalize_ntpath(wbuf), -1, 0, 0, NULL, NULL);
-	if (len) {
-		*ptag = b->ReparseTag;
-		return len - 1;
-	}
-	errno = ERANGE;
-	return -1;
+	if ((*plen =
+	     xwcstoutf(tmpbuf, normalize_ntpath(wbuf), MAX_LONG_PATH)) <  0)
+		return -1;
+	return 0;
 }
 
 int readlink(const char *path, char *buf, size_t bufsiz)
 {
-	WCHAR wpath[MAX_LONG_PATH], *wbuf;
-	REPARSE_DATA_BUFFER *b = alloca(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+	WCHAR wpath[MAX_LONG_PATH];
 	char tmpbuf[MAX_LONG_PATH];
 	int len;
+	DWORD tag;
 
 	if (xutftowcs_long_path(wpath, path) < 0)
 		return -1;
 
-	if (get_reparse_point(wpath, b, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &wbuf) < 0)
-		return -1;
-	if (wbuf == NULL) {
-		errno = EINVAL;
+	if (readlink_1(wpath, tmpbuf, &len, &tag) < 0)
 		return -1;
-	}
 
 	/*
 	 * Adapt to strange readlink() API: Copy up to bufsiz *bytes*, potentially
@@ -3061,8 +3041,6 @@ int readlink(const char *path, char *buf, size_t bufsiz)
 	 * so convert to a (hopefully large enough) temporary buffer, then memcpy
 	 * the requested number of bytes (including '\0' for robustness).
 	 */
-	if ((len = xwcstoutf(tmpbuf, normalize_ntpath(wbuf), MAX_LONG_PATH)) < 0)
-		return -1;
 	memcpy(buf, tmpbuf, min(bufsiz, len + 1));
 	return min(bufsiz, len);
 }
```

billziss-gh replied:
I think this will work in the general case. However please note that with the fixup, `lstat` will fail with `EINVAL` in the case of a reparse point that is neither a symlink nor a mountpoint. (That was the primary reason that I refactored into the separate `get_reparse_point` in my earlier commit.)

I any case I am happy to include this fixup and resubmit. I will apply your fixup as a patch on top of my commit and attribute the changes in the commit message.

dscho replied:
> However please note that with the fixup, `lstat` will fail with `EINVAL` in the case of a reparse point that is neither a symlink nor a mountpoint. (That was the primary reason that I refactored into the separate `get_reparse_point` in my earlier commit.)

What could it be, then?

If you care very deeply about this, you can very easily add a parameter to `readlink_1()`, say, `fail_on_unknown_tag`, and only do the `EINVAL` thing if that parameter is non-zero, otherwise do the `MAX_LONG_PATH` thing.

billziss-gh commented:
I have pushed a new commit with your fixup applied. I have also added the `fail_on_unknown_tag` change.

billziss-gh commented:
@dscho glad to be of help and thank you for merging this in.

billziss-gh commented:
I just updated to the latest git-for-windows and I am still having the original problem with symbolic links. Perhaps the change in the PR was inadvertently left out?

Git-for-windows repo cloned in WSL:

```console
$ uname -a
Linux xps 4.4.0-19041-Microsoft #1-Microsoft Fri Dec 06 14:06:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
$ git clone https://github.com/git-for-windows/git.git
Cloning into 'git'...
remote: Enumerating objects: 441232, done.
remote: Total 441232 (delta 0), reused 0 (delta 0), pack-reused 441232
Receiving objects: 100% (441232/441232), 176.09 MiB | 9.64 MiB/s, done.
Resolving deltas: 100% (319794/319794), done.
Updating files: 100% (3787/3787), done.
$ cd git
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Latest status on same repo from Powershell using git-for-windows 2.27.0.windows.1:

```console
> git --version
git version 2.27.0.windows.1
> git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   RelNotes

no changes added to commit (use "git add" and/or "git commit -a")
```

dscho commented:
> Perhaps the change in the PR was inadvertently left out?

No, it was not: https://github.com/git-for-windows/git/blob/v2.27.0.windows.1/compat/mingw.c#L996-L1003

But yes, I can verify here what you see. This is strange, and maybe it is an unfortunate side effect of other changes that made it into v2.27.0? I _did_ test this before merging and it worked over here, but of course the merged branch was based on v2.26.2.

billziss-gh commented:
I am currently working on something else, but will try to have a look. Unfortunately I had some problems with the gdb that ships with the Git for Windows SDK.

billziss-gh commented:
@dscho I synced my fork and tried a build. The resulting `git-status.exe` appears to work correctly:

From git-bash:

```console
$ uname -a
MINGW64_NT-10.0-19041 xps 3.1.4-340.x86_64 2020-05-19 12:55 UTC x86_64 Msys
$ ./git.exe --version
git version 2.27.0.windows.1.2.gce436fdb35.dirty
$ ./git-show.exe
commit ce436fdb35e7c90706b56d8b5debbb1edd5bb65c (HEAD -> master, upstream/master, origin/master, origin/HEAD)
Merge: 907ab1011d 8422b5e207
Author: Johannes Schindelin <Johannes.Schindelin@gmx.de>
Date:   Thu Jun 4 23:50:17 2020 +0200

    Merge pull request #2665 from dscho/fix-msvc-build-g4w

    msvc: fix "REG_STARTEND" issue

$ ./git-status.exe
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

From git-bash with git 2.27.0.windows.1:

```console
$ uname -a
MINGW64_NT-10.0-19041 xps 3.1.4-340.x86_64 2020-05-19 12:55 UTC x86_64 Msys
$ /c/Program\ Files/Git/cmd/git.exe --version
git version 2.27.0.windows.1
$ /c/Program\ Files/Git/cmd/git.exe status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   RelNotes

no changes added to commit (use "git add" and/or "git commit -a")
```

From WSL1:

```console
$ uname -a
Linux xps 4.4.0-19041-Microsoft #1-Microsoft Fri Dec 06 14:06:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
$ git --version
git version 2.25.1
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

So this shows that the build from latest master and WSL1 agree on the repo status and they both disagree with git 2.27.0.windows.1.

billziss-gh commented:
The mystery deepens...

Because the git 2.27.0.windows.1 release does not have any symbols I ended up decompiling `git-status.exe` using Ghidra. Although symbols were missing it was relatively easy to find the right spot by looking for references to `DeviceIoControl`. Luckily there was a single location where this API is used.

Turns out that the surrounding code looks like the `readlink_1` function from last week. This of course means that git  2.27.0.windows.1 **does include** the fix from this PR. Which makes it hard to understand where the problem is.

Do you by any chance have the symbols from the 2.27.0.windows.1 build? Debugging the release build would give us the real answer.

I include the Ghidra code for completeness:

```C
int readlink_1(LPCWSTR wpath,int fail_on_unknown_tag,char *tmpbuf,int *plen,DWORD *ptag)
{
  BOOL BVar1;
  int iVar2;
  DWORD DVar3;
  longlong lVar4;
  HANDLE handle;
  undefined8 uVar5;
  int *piVar6;
  undefined8 auStack160 [5];
  DWORD *pDStack120;
  DWORD aDStack112 [2];
  LPDWORD pDStack104;
  LPOVERLAPPED p_Stack96;
  DWORD aDStack88 [2];
  code *local_50;
  undefined2 auStack72 [4];
  
  local_50 = __stack_chk_guard_exref;
  auStack160[0] = 0x62a20d;
  lVar4 = FUN_006571b0();
  lVar4 = -lVar4;
  *(LPDWORD *)((longlong)&pDStack104 + lVar4) = (LPDWORD)0x0;
  *(DWORD *)((longlong)aDStack112 + lVar4) = 0x2200000;
  *(undefined4 *)((longlong)&pDStack120 + lVar4) = 3;
  *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a23f;
  handle = CreateFileW(wpath,0,7,(LPSECURITY_ATTRIBUTES)0x0,
                       *(DWORD *)((longlong)&pDStack120 + lVar4),
                       *(DWORD *)((longlong)aDStack112 + lVar4),
                       *(LPDWORD *)((longlong)&pDStack104 + lVar4));
  if (handle == (HANDLE)0xffffffffffffffff) {
    *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a35d;
    DVar3 = GetLastError(*(undefined *)((longlong)auStack160 + lVar4));
    *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a366;
    piVar6 = _errno(*(undefined *)((longlong)auStack160 + lVar4));
    *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a371;
    iVar2 = FUN_00628180((ulonglong)DVar3);
    *piVar6 = iVar2;
    iVar2 = -1;
  }
  else {
    *(DWORD **)((longlong)&pDStack120 + lVar4) = (DWORD *)((longlong)aDStack88 + lVar4);
    *(LPDWORD *)((longlong)&pDStack104 + lVar4) = aDStack88 + 1;
    *(LPOVERLAPPED *)((longlong)&p_Stack96 + lVar4) = (LPOVERLAPPED)0x0;
    *(DWORD *)((longlong)aDStack112 + lVar4) = 0x4000;
    *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a27f;
    BVar1 = DeviceIoControl(handle,0x900a8,(LPVOID)0x0,0,*(DWORD **)((longlong)&pDStack120 +lVar4),
                            *(DWORD *)((longlong)aDStack112 + lVar4),
                            *(LPDWORD *)((longlong)&pDStack104 + lVar4),
                            *(LPOVERLAPPED *)((longlong)&p_Stack96 + lVar4));
    if (BVar1 == 0) {
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a32e;
      DVar3 = GetLastError(*(undefined *)((longlong)auStack160 + lVar4));
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a337;
      piVar6 = _errno(*(undefined *)((longlong)auStack160 + lVar4));
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a342;
      iVar2 = FUN_00628180((ulonglong)DVar3);
      *piVar6 = iVar2;
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a34d;
      CloseHandle(handle,*(undefined *)((longlong)auStack160 + lVar4));
      iVar2 = -1;
    }
    else {
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a290;
      CloseHandle(handle,*(undefined *)((longlong)auStack160 + lVar4));
      DVar3 = *(DWORD *)((longlong)aDStack88 + lVar4);
      *ptag = DVar3;
      if (DVar3 == 0xa0000003) {
        *(undefined2 *)
         ((longlong)auStack72 +
         (ulonglong)*(ushort *)((longlong)&local_50 + lVar4 + 2) +
         (ulonglong)*(ushort *)((longlong)&local_50 + lVar4) + lVar4) = 0;
      }
      else {
        if (DVar3 != 0xa000000c) {
          if (fail_on_unknown_tag == 0) {
            *plen = 0x1000;
            iVar2 = 0;
          }
          else {
            *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a383;
            piVar6 = _errno(*(undefined *)((longlong)auStack160 + lVar4));
            *piVar6 = 0x16;
            iVar2 = -1;
          }
          goto LAB_0062a2d3;
        }
        *(undefined2 *)
         ((longlong)auStack72 +
         (ulonglong)*(ushort *)((longlong)&local_50 + lVar4 + 2) +
         (ulonglong)*(ushort *)((longlong)&local_50 + lVar4) + lVar4 + 4) = 0;
      }
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a2bd;
      uVar5 = FUN_00627340();
      *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a2ce;
      iVar2 = FUN_0062a140(tmpbuf,uVar5,0x1000);
      *plen = iVar2;
      iVar2 = iVar2 >> 0x1f;
    }
  }
LAB_0062a2d3:
  if (local_50 == __stack_chk_guard_exref) {
    return iVar2;
  }
                    /* WARNING: Subroutine does not return */
  *(undefined8 *)((longlong)auStack160 + lVar4) = 0x62a398;
  __stack_chk_fail();
}
```

dscho commented:
> Do you by any chance have the symbols from the 2.27.0.windows.1 build?

Yes: https://github.com/git-for-windows/git/releases/download/v2.27.0.windows.1/pdbs-for-git-64-bit-2.27.0.1.907ab1011d-1.zip

Of course, those are `.pdb` files, which GDB won't understand... But maybe you did not want to use GDB at all?

dscho commented:
I tried to run `git status` in WinDbg (don't run `git-status.exe`, it is just the [Git wrapper](https://github.com/git-for-windows/MINGW-packages/blob/master/mingw-w64-git/git-wrapper.c), not the actual built-in). It does look as if `readlink_1()` is called _only_ when locking the index file:

```
0:000> k
 # Child-SP          RetAddr           Call Site
00 00000000`00bfc188 00000000`0027d6bc git!readlink_1 [compat\mingw.c @ 2981] 
01 00000000`00bfc190 00000000`00230eb0 git!readlink+0x9c [compat\mingw.c @ 3044] 
02 00000000`00bff220 00000000`001936a5 git!strbuf_readlink+0x70 [strbuf.c @ 576] 
03 00000000`00bff290 00000000`00193a88 git!lock_file+0xf5 [lockfile.c @ 54] 
04 00000000`00bff320 00000000`001f8976 git!hold_lock_file_for_update_timeout_mode+0x128 [lockfile.c @ 111] 
05 00000000`00bff3c0 00000000`00076e93 git!repo_hold_locked_index+0x26 [repository.c @ 281] 
06 00000000`00bff400 00000000`000517bc git!cmd_status+0x393 [builtin\commit.c @ 1454] 
07 00000000`00bff4a0 00000000`00052aa0 git!handle_builtin+0x22c [git.c @ 448] 
08 00000000`00bff570 00000000`002a8578 git!cmd_main+0x1b0 [git.c @ 777] 
09 00000000`00bff6e0 00000000`0027f7e9 git!main+0x78 [common-main.c @ 54] 
0a 00000000`00bff720 00000000`000513e4 git!wmain+0x299 [compat\mingw.c @ 3733] 
0b 00000000`00bff7a0 00000000`00051516 git!__tmainCRTStartup+0x254 [D:\mingwbuild\mingw-w64-crt-git\src\mingw-w64\mingw-w64-crt\crt\crtexe.c @ 337] 
0c 00000000`00bff870 00007ffd`28f86fd4 git!mainCRTStartup+0x16 [D:\mingwbuild\mingw-w64-crt-git\src\mingw-w64\mingw-w64-crt\crt\crtexe.c @ 221] 
0d 00000000`00bff8a0 00007ffd`2a17cec1 KERNEL32!BaseThreadInitThunk+0x14
0e 00000000`00bff8d0 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

I found [the WinDbg Getting Started](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/getting-started-with-windbg) very helpful. Essentially, I set a breakpoint (`bu git!readlink_1`) and started/continued the process (`g`) and inspected the stack trace (`k`).

billziss-gh commented:
Perfect if you have PDBs. I am very familiar with WinDbg as I use it for kernel debugging. I will have a look later today. (It is morning here in Seattle.)

billziss-gh commented:
Here are my few findings so far:

- This experiment was performed from git-bash that ships with git-sdk.
    - If I run `/c/Program\ Files/Git/mingw64/bin/git.exe status` the problem persists (i.e. symlinks are reported as modified).
    - If I copy the `/c/Program\ Files/Git/mingw64/bin/git.exe` executable locally and then execute `./git.exe status` then the problem goes away! (i.e. are symlinks are **not** reported as modified).
- This experiment was performed from git-bash that ships with Git-for-windows. I have also used the git-sdk to build the latest master of Git-for-Windows:
    - If I run `/c/Program\ Files/Git/mingw64/bin/git.exe status` the problem persists.
    - If I run `./git.exe status` then the problem goes away!
    - **If I then copy the newly built `git.exe` into `/c/Program\ Files/Git/mingw64/bin/git.exe` and then run `/c/Program\ Files/Git/mingw64/bin/git.exe status` the problem persists.**

So the results seem to be affected by which mingw environment is picked up by git when it starts running.

It is exceedingly hard to use WinDbg as the PDB symbols appear to be inadequate and perhaps even corrupted. (Perhaps because they were not created with Microsoft dev tools?) And trying with GDB that is bundled with git-for-windows fails with `Cannot insert breakpoint 1. Cannot access memory at address 0x426b00`.

In any case in my experiments with WinDbg (against git version 2.27.0.windows.1) I noticed the following:

- `mingw_lstat` is never called for files in the work tree. It is as if Git decides to optimize these lstat calls away, because it does not expect that files have changed.
- `mingw_lstat` is only ever called for files in `.git`, such as `.git/index`, etc.

dscho commented:
@billziss-gh oh you know, I think I understand the problem. Do you have FSCache enabled? If so, I think `fcache_lstat()` (I _think_ that's the name of the function in `compat/win32/fscache.c`) needs to be patched, too.

billziss-gh commented:
Good catch.

I believe a change should be made in [`fseentry_create_entry`](https://github.com/git-for-windows/git/blob/a98ad18cd6176cb909d409086f138c25f7d1c68e/compat/win32/fscache.c#L223) to report the correct size. (A global search for `MAX_LONG_PATH` did not reveal any other place where that constant is being used to set `st_size`, but please let me know if there might be another place.)

I attempted a patch reusing the computation of `buf` [here](https://github.com/git-for-windows/git/blob/a98ad18cd6176cb909d409086f138c25f7d1c68e/compat/win32/fscache.c#L206-L217), but `buf` appears to be computed wrong. This may mean that we have to pass `pattern` from `fsentry_create_list` into `fsentry_create_entry`.

```patch
diff --git a/compat/win32/fscache.c b/compat/win32/fscache.c
index 5d929a1fc3..a4c55899ad 100644
--- a/compat/win32/fscache.c
+++ b/compat/win32/fscache.c
@@ -185,6 +185,7 @@ static struct fsentry *fseentry_create_entry(struct fscache *cache,
 	char buf[MAX_PATH * 3];
 	int len;
 	struct fsentry *fse;
+	int link_len = MAX_LONG_PATH;
 
 	len = xwcstoutfn(buf, ARRAY_SIZE(buf), fdata->FileName, fdata->FileNameLength / sizeof(wchar_t));
 
@@ -195,7 +196,12 @@ static struct fsentry *fseentry_create_entry(struct fscache *cache,
 		fdata->EaSize : 0;
 
 	/*
-	 * On certain Windows versions, host directories mapped into
+	 * We compute a path that can be opened for 2 reasons:
+	 *
+	 * 1. For symbolic links we must calculate the link length so
+	 * that we can report it in st_size.
+	 *
+	 * 2. On certain Windows versions, host directories mapped into
 	 * Windows Containers ("Volumes", see https://docs.docker.com/storage/volumes/)
 	 * look like symbolic links, but their targets are paths that
 	 * are valid only in kernel mode.
@@ -204,8 +210,7 @@ static struct fsentry *fseentry_create_entry(struct fscache *cache,
 	 * telling Git that these are *not* symbolic links.
 	 */
 	if (fse->reparse_tag == IO_REPARSE_TAG_SYMLINK &&
-	    sizeof(buf) > (list ? list->len + 1 : 0) + fse->len + 1 &&
-	    is_inside_windows_container()) {
+	    sizeof(buf) > (list ? list->len + 1 : 0) + fse->len + 1) {
 		size_t off = 0;
 		if (list) {
 			memcpy(buf, list->dirent.d_name, list->len);
@@ -214,13 +219,18 @@ static struct fsentry *fseentry_create_entry(struct fscache *cache,
 		}
 		memcpy(buf + off, fse->dirent.d_name, fse->len);
 		buf[off + fse->len] = '\0';
+
+		char tmpbuf[MAX_LONG_PATH];
+		link_len = readlink(buf, tmpbuf, sizeof tmpbuf);
+		if (link_len == -1)
+			link_len = MAX_LONG_PATH;
 	}
 
 	fse->st_mode = file_attr_to_st_mode(fdata->FileAttributes,
 					    fdata->EaSize, buf);
 	fse->dirent.d_type = S_ISREG(fse->st_mode) ? DT_REG :
 			S_ISDIR(fse->st_mode) ? DT_DIR : DT_LNK;
-	fse->u.s.st_size = S_ISLNK(fse->st_mode) ? MAX_LONG_PATH :
+	fse->u.s.st_size = S_ISLNK(fse->st_mode) ? link_len :
 			fdata->EndOfFile.LowPart |
 			(((off_t)fdata->EndOfFile.HighPart) << 32);
 	filetime_to_timespec((FILETIME *)&(fdata->LastAccessTime),
```

Running from git-bash (with an additional `fprintf(stderr)` not included in the patch above):
```console
billziss@xps:~/Projects/ext/git-for-windows [master]$ ./git.exe config core.fscache
true
billziss@xps:~/Projects/ext/git-for-windows [master]$ ./git.exe status
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   compat/win32/fscache.c

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   compat/win32/fscache.c

billziss@xps:~/Projects/ext/git-for-windows [master]$ cd compat/
billziss@xps:~/Projects/ext/git-for-windows/compat [master]$ ../git.exe status
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
readlink("/RelNotes") = 4096
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   win32/fscache.c

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   win32/fscache.c
```

dscho commented:
Give me five minutes, I think I have it working.

dscho commented:
Here it is: https://github.com/git-for-windows/git/pull/2673. Can I ask you to test and review this PR?

billziss-gh commented:
Will do. Thanks!

ariccio commented:
I'm a totally uninvolved party, but if I may, I can impart a tiny bit of knowledge as the developer of altWinDirStat. Indeed, when iterating over a directory, `FindFirstFile` will 'cache' results as it is really just translating the data from the `NtQueryDirectoryFile` call. This is also why the API really doesn't scale well at all, since using the WIN32 API means it will take a separate call and copy to enumerate each file on the system. This can easily saturate a single core on initializing and copying structs alone when the operation is no longer IO bound.

The `NtQueryDirectoryFile` API is (IMO) a vastly better way to iterate over a directory tree, but can produce surprising results for users as you're no longer working with familiar DOS paths. It's also, IMO, not absolutely 100% clearly documented enough for me to feel like there's a truly safe UB and surprise free way to parse the raw buffer. Which, personally, I would like before ditching the ultra-stable WIN32 API. I think this is a known issue, and I believe I discussed it very briefly in a private communication with Alex Ionescu (who I'm mentioning here mainly because he's the Windows Internals god).

In my experience `FindFirstFileExA` with `FIND_FIRST_EX_LARGE_FETCH` doesn't usually make an improvement. [I profiled it several years ago](https://ariccio.com/2015/01/13/how-fast-are-findfirstfilefindfirstfileex-and-cfilefind-actually/), and I found it made things *slower* for some reason.

I don't think this knowledge will do any good in this issue here, but I hope someone finds it interesting or useful.

dscho commented:
@ariccio thank you for sharing your insights!

jmiserez commented:
Hi @dscho.

You seem to be working on a bug fix related to this in combination to symlinks and git status, so I think this is relevant.

This PR (https://github.com/git-for-windows/git/pull/2637) seems to have introduced a performance regression which was not present in 2.26.1, but is present in 2.27.0 onwards, namely extremely slow performance (~126s) when running `git status --ignored` on a repository with deep folder structures (97 folders deep) containing symlinks (specifically pnpm folders). `git status` without `--ignored` works fine (~0.17s), as the folders in question are then ignored.

**Why do I believe this MR is the culprit?**
In my tests, I've reproduced the issue when running Git for Windows 2.27.0 and later on Windows 10 (on a Windows filesystem), but not when running the identical Linux/Ubuntu version on the same Windows filesystem (via WSL1). The only relevant PR I can find is this one, so I believe it to be the root cause. The issue is present for a large number of users in our organization across many different Windows 10 machines, but is never present when using Git for Windows 2.26.1 or earlier on the same machines.

I'm sorry that I don't have a specific repository/project to demonstrate this issue, but it does happen across tens of users and several repositories at our organization. We have since been able to reduce the depth of our pnpm folder structures to a more reasonable number (~15) by changing our project setup, but the issue of slower performance vs. older versions of Git for Windows and the Linux git clients remains and even with the lower depth is still 5x slower than before. It's pretty hard to come up with a public repository that replicates the issue cleanly enough to demonstrate, but maybe this information gives you an idea of what could be wrong.

### This is with the newest Git and after our reorganization, so symlinked directories only ~15 folders deep:
**Git for Windows (git version 2.32.0.windows.1) via Git Bash:**
```
$ GIT_TRACE_PERFORMANCE=1 git status
11:09:42.130149 read-cache.c:2381       performance: 0.003948300 s:  read cache .git/index
11:09:42.236657 preload-index.c:159     performance: 0.104057800 s:   preload index
11:09:42.236657 read-cache.c:1682       performance: 0.104883000 s:  refresh index
11:09:42.239657 diff-lib.c:262          performance: 0.000610700 s:  diff-files
11:09:42.243656 unpack-trees.c:1687     performance: 0.000050000 s:    traverse_trees
11:09:42.244658 unpack-trees.c:413      performance: 0.000002200 s:    check_updates
11:09:42.244658 unpack-trees.c:1773     performance: 0.000556000 s:   unpack_trees
11:09:42.244658 diff-lib.c:606          performance: 0.000782500 s:  diff-index
11:09:42.246658 name-hash.c:607         performance: 0.002159500 s:  initialize name hash
11:09:42.296657 trace.c:487             performance: 0.173544800 s: git command: 'C:\Users\jms\AppData\Local\Programs\Git\mingw64\bin\git.exe' status
```

```
$ GIT_TRACE_PERFORMANCE=1 git status --ignored
11:10:20.932380 read-cache.c:2381       performance: 0.003731500 s:  read cache .git/index
11:10:21.027992 preload-index.c:159     performance: 0.092727600 s:   preload index
11:10:21.027992 read-cache.c:1682       performance: 0.093447300 s:  refresh index
11:10:21.029991 diff-lib.c:262          performance: 0.000537100 s:  diff-files
11:10:21.033992 unpack-trees.c:1687     performance: 0.000043300 s:    traverse_trees
11:10:21.034992 unpack-trees.c:413      performance: 0.000001800 s:    check_updates
11:10:21.034992 unpack-trees.c:1773     performance: 0.000408800 s:   unpack_trees
11:10:21.034992 diff-lib.c:606          performance: 0.000588000 s:  diff-index
11:10:21.036991 name-hash.c:607         performance: 0.001704700 s:  initialize name hash
11:11:58.091387 trace.c:487             performance: 10.611363200 s: git command: 'C:\Users\jms\AppData\Local\Programs\Git\mingw64\bin\git.exe' status --ignored
```

**Git on Ubuntu (git version 2.32.0) via WSL on same filesystem:**

```
GIT_TRACE_PERFORMANCE=1 git status
11:13:18.662336 read-cache.c:2368       performance: 0.006310800 s:  read cache .git/index
11:13:19.827091 preload-index.c:154     performance: 1.163495000 s:   preload index
11:13:19.827939 read-cache.c:1670       performance: 1.164346800 s:  refresh index
11:13:19.830386 diff-lib.c:262          performance: 0.000567900 s:  diff-files
11:13:19.837891 unpack-trees.c:1685     performance: 0.000040100 s:    traverse_trees
11:13:19.838186 unpack-trees.c:413      performance: 0.000002500 s:    check_updates
11:13:19.838202 unpack-trees.c:1771     performance: 0.000802800 s:   unpack_trees
11:13:19.838209 diff-lib.c:606          performance: 0.001186300 s:  diff-index
11:13:19.842639 name-hash.c:607         performance: 0.003828200 s:  initialize name hash
11:13:20.408649 trace.c:487             performance: 1.756890900 s: git command: git status
```

```
GIT_TRACE_PERFORMANCE=1 git status --ignored
11:13:46.147775 read-cache.c:2368       performance: 0.006347700 s:  read cache .git/index
11:13:47.331391 preload-index.c:154     performance: 1.182246300 s:   preload index
11:13:47.332205 read-cache.c:1670       performance: 1.183065000 s:  refresh index
11:13:47.334850 diff-lib.c:262          performance: 0.000583400 s:  diff-files
11:13:47.343517 unpack-trees.c:1685     performance: 0.000044200 s:    traverse_trees
11:13:47.343805 unpack-trees.c:413      performance: 0.000002500 s:    check_updates
11:13:47.343817 unpack-trees.c:1771     performance: 0.000844500 s:   unpack_trees
11:13:47.343822 diff-lib.c:606          performance: 0.001266200 s:  diff-index
11:13:47.348746 name-hash.c:607         performance: 0.004165500 s:  initialize name hash
11:13:48.780322 trace.c:487             performance: 2.643303900 s: git command: git status --ignore
```

### With an older Git 2.29.0 and before our reorganization, symlinked directories ~100 folders deep:
Here the numbers were much worse and there was also an additional "dir.c:2824" entry just before the last line, where the bulk of the time was spent. This isn't shown in the trace anymore with 2.32.0, but I have some old traces from then:

**Git for Windows 2.29.0 shows this additional "dir.c:2824" entry:**
GIT_TRACE_PERFORMANCE=1 git status  ```

23:02:02.522268 read-cache.c:2310       performance: 0.002348200 s:  read cache .git/index  
23:02:02.639267 preload-index.c:152     performance: 0.112831400 s:   preload index  
23:02:02.639267 read-cache.c:1621       performance: 0.113058400 s:  refresh index  
23:02:02.644267 diff-lib.c:249          performance: 0.000105700 s:  diff-files  
23:02:02.663268 unpack-trees.c:1659     performance: 0.000052000 s:    traverse_trees  
23:02:02.663268 unpack-trees.c:411      performance: 0.000000800 s:    check_updates  
23:02:02.663268 unpack-trees.c:1746     performance: 0.000188800 s:   unpack_trees  
23:02:02.663268 diff-lib.c:529          performance: 0.000737200 s:  diff-index  
23:02:02.665269 name-hash.c:600         performance: 0.001713600 s:   initialize name hash 
23:02:02.683267 dir.c:2824              performance: 0.019035400 s:  read directory  
23:02:02.689269 trace.c:487             performance: 0.176100700 s: git command: 'C:\Program Files\Git\mingw64\bin\git.exe' status
```

```
GIT_TRACE_PERFORMANCE=1 git status  
22:57:15.710535 read-cache.c:2310       performance: 0.001907600 s:  read cache .git/index  
22:57:15.843535 preload-index.c:152     performance: 0.129294400 s:   preload index  
22:57:15.845533 read-cache.c:1621       performance: 0.131907700 s:  refresh index  
22:57:15.850534 diff-lib.c:249          performance: 0.000132900 s:  diff-files  
22:57:15.862535 unpack-trees.c:1659     performance: 0.000046000 s:    traverse_trees  
22:57:15.862535 unpack-trees.c:411      performance: 0.000000300 s:    check_updates  
22:57:15.862535 unpack-trees.c:1746     performance: 0.000186100 s:   unpack_trees  
22:57:15.862535 diff-lib.c:529          performance: 0.000291100 s:  diff-index  
22:57:15.865534 name-hash.c:600         performance: 0.002165100 s:   initialize name hash 
22:59:21.922352 dir.c:2824              performance: 126.057996100 s:  read directory   
22:59:21.937398 trace.c:487             performance: 126.237585500 s: git command: 'C:\Program Files\Git\mingw64\bin\git.exe' status --ignored
```
